sympy/combinatorics/perm_groups.py

```diff
@@ -1702,6 +1705,68 @@ def is_abelian(self):
                     return False
         return True
 
+    def abelian_inv(self):
+        """
+        Returns the abelian invariants or primary decomposition of the commutator
+        factor group of the given group ``G`` as suggested in ([14], p. 542).
+        These are given as a list of prime-powers or zeroes and describe the
+        structure of ``G/G'`` as a direct product of cyclic groups of prime power order.
+
+        Examples
+        ========
+
+        >>> from sympy.combinatorics import Permutation
+        >>> from sympy.combinatorics.perm_groups import PermutationGroup
+        >>> a = Permutation([0, 2, 1])
+        >>> b = Permutation([1, 0, 2])
+        >>> G = PermutationGroup([a, b])
+        >>> G.abelian_inv()
+        [2]
+        >>> from sympy.combinatorics.named_groups import CyclicGroup
+        >>> G = CyclicGroup(7)
+        >>> G.abelian_inv()
+        [7]
+
+        """
+        if self.is_trivial:
+            return []
+        gns = self.generators
+        inv = []
+        der = self.derived_subgroup()
+        G = self
+
+        for p in primefactors(G.order()):
+            ranks = []
+            r = 2
+            while(r != 1):
+                H = der
+                for g in gns:
+                    elm = g**p
+                    if not H.contains(elm):
+                        for i in range(len(H)):
+                            H = PermutationGroup([H[i], elm])
```

Comment:
It looks `H` will be repeatedly redefined (and a single `H[i]` included). Perhaps this should be something like

    H = PermutationGroup(H.generators + [elm])

Reply:
yes, that's right.